File: solutions/PYTHON/missing_number.py
# 
# Python function to find the missing number in a sequence of integers.
# 
# Author: Uthman Dev
# 
# Description: This function calculates the expected sum of a sequence of
#              consecutive integers from 1 up to the length of the array plus one.
#              By subtracting the actual sum of the given array from the expected sum,
#              it determines the missing number in the sequence.
#

def missing_number(array):
    n = len(array) + 1
    expected_sum = n * (n + 1) // 2
    actual_sum = sum(array)
    missing = expected_sum - actual_sum
    return missing

# Example usage
array = [1, 2, 3, 4, 6]
missing = missing_number(array)
print("Missing number:", missing)  # Expected output: Missing number, e.g., 5

File: solutions/PHP/missing_number.php
<?php
declare(strict_types=1);
/**
 * PHP class to find the missing number in a sequence.
 *
 * Author: Uthman Dev
 *
 * Description: This class calculates the missing number in a sequence of integers
 * where only one number is missing. It uses the sum formula for a range to find
 * the missing value and then adds it back to the array.
 */
class MissingNumberFinder
{
  private $numbers; // The input array of numbers

  /**
   * Constructor to initialize the input array
   *
   * @param array $numbers - The array of numbers in the sequence
   */
  public function __construct(array $numbers)
  {
    $this->numbers = $numbers;
  }

  /**
   * Method to calculate, add, and display the missing number in the sequence
   *
   * @return void
   */
  public function findAndAddMissingNumber(): void
  {
    // Calculate the sum of numbers in the array
    $actualSum = array_sum($this->numbers);

    // Find the maximum expected number in the sequence
    $maxNumber = max($this->numbers);

    // Calculate the expected sum for a full sequence from 1 to maxNumber
    $expectedSum = ($maxNumber * ($maxNumber + 1)) / 2;

    // Calculate the missing number as the difference between expected and actual sums
    $missingNumber = $expectedSum - $actualSum;

    // Add the missing number to the array
    $this->numbers[] = $missingNumber;

    // Sort the array to keep the sequence in order
    sort($this->numbers);

    // Output the result
    echo "The missing number is: $missingNumber\n";
    echo "Updated array with the missing number added: " . implode(", ", $this->numbers) . "\n";
  }
}

// Example usage
$numbers = [1, 2, 4, 5]; // Sequence with missing number 3
$missingFinder = new MissingNumberFinder($numbers);
$missingFinder->findAndAddMissingNumber();
?>
